dynet/cudnn-ops.cc

```diff
@@ -0,0 +1,227 @@
+    throw std::runtime_error("dynet::CudnnConvOp::mempool_ not set");
+  if (h_odd || w_odd) {
+    unsigned new_XH = XH + h_odd;
+    unsigned new_XW = XW + w_odd;
+    DYNET_ASSERT(padded_x.d[0] == new_XH, "Tensor input_padded must have been padded");
+    DYNET_ASSERT(padded_x.d[1] == new_XW, "Tensor input_padded must have been padded");
+    x = &padded_x;
+    XH = new_XH;
+    XW = new_XW;
+    if (i == 0)
+      dx_ptr = mempool_->allocate(sizeof(float) * new_XW * new_XH * XC * XN);
+  }
+  // here we could reuse the descriptor we created for forward, because 
+  // they share the same size
+  float alpha = 1.f, beta = 0.f;
+  if (i == 1) {
+    CUDNN_CHECK(cudnnGetConvolutionBackwardFilterAlgorithm(dev.cudnnHandle,
+                x_desc_, y_desc_, conv_desc_, filter_desc_, 
+                CUDNN_CONVOLUTION_BWD_FILTER_PREFER_FASTEST, 
+                workspace_size_limit_bytes, &bwd_f_algo_));
+    CUDNN_CHECK(cudnnGetConvolutionBackwardFilterWorkspaceSize(dev.cudnnHandle,
+                x_desc_, y_desc_, conv_desc_, filter_desc_,
+                bwd_f_algo_, &workspace_bwd_filter_size_));
+    // allocate space for backward compute
+    if (bwd_filter_workspace == NULL) {
+      bwd_filter_workspace = mempool_->allocate(sizeof(float) * workspace_bwd_filter_size_);
+    }
+    CUDNN_CHECK(cudnnConvolutionBackwardFilter(dev.cudnnHandle,
+                &alpha, x_desc_, x->v,
+                y_desc_, dy->v,
+                conv_desc_, bwd_f_algo_, bwd_filter_workspace, workspace_bwd_filter_size_,
+                &beta, filter_desc_, dxi->v));
+  } else if (i == 0) {            
+    CUDNN_CHECK(cudnnGetConvolutionBackwardDataAlgorithm(dev.cudnnHandle,
+                filter_desc_, y_desc_, conv_desc_, x_desc_,
+                CUDNN_CONVOLUTION_BWD_DATA_PREFER_FASTEST,
+                workspace_size_limit_bytes, &bwd_d_algo_));
+    CUDNN_CHECK(cudnnGetConvolutionBackwardDataWorkspaceSize(dev.cudnnHandle,
+                filter_desc_, y_desc_, conv_desc_, x_desc_,
+                bwd_d_algo_, &workspace_bwd_data_size_));
+    if (bwd_data_workspace == NULL) {
+      bwd_data_workspace = mempool_->allocate(sizeof(float) * workspace_bwd_data_size_);
+      }
+    if (h_odd || w_odd) {
+      CUDNN_CHECK(cudnnConvolutionBackwardData(dev.cudnnHandle,
+                  &alpha, filter_desc_, filter->v,
+                  y_desc_, dy->v,
+                  conv_desc_, bwd_d_algo_, bwd_data_workspace, workspace_bwd_data_size_,
+                  &beta, x_desc_, dx_ptr));
+      Tensor padded_dx = Tensor(Dim({XH, XW, XC}, XN), static_cast<float*>(dx_ptr), xs[0]->device, DeviceMempool::FXS);
+        
+      //Eigen::array<int, 4> offsets = {0, 0, 0, 0};
+      //Eigen::array<int, 4> extents = {static_cast<int>(XH), static_cast<int>(XW), static_cast<int>(XC), static_cast<int>(XN)};
+      //TODO this functio cannot be linked
+      //dxi->tb<3>().device(*dev.edevice) = padded_dx.tb<3>().slice(offsets, extents);
```

Comment:
Here I tried to use Eigen slice() function to do negative padding in place of my customized functions but ended up with a linking error. Can someone familiar with Eigen GPU operations take a look?